File: netlify/functions/check-immatriculations.mjs
/**
 * Fonction Netlify planifiée — tourne tous les lundis à 8h
 * Interroge l'API data.gouv, stocke les nouvelles immatriculations dans Supabase
 * et envoie une alerte email via Resend.
 */

import { schedule } from "@netlify/functions";
import { createClient } from "@supabase/supabase-js";
import { Resend } from "resend";

// --- Configuration ---

const API_URL = "https://recherche-entreprises.api.gouv.fr/search";

const APE_CODES = [
  "56.10A", // Restauration traditionnelle
  "56.10C", // Restauration rapide
  "55.10Z", // Hôtels et hébergement similaire
  "56.30Z", // Débits de boissons
  "10.71C", // Boulangerie-pâtisserie
  "10.71D", // Pâtisserie
  "47.22Z", // Boucherie-charcuterie
  "47.24Z", // Commerce pain/pâtisserie
  "56.21Z", // Services traiteurs
];

const APE_LABELS = {
  "56.10A": "Restauration traditionnelle",
  "56.10C": "Restauration rapide",
  "55.10Z": "Hôtels et hébergement",
  "56.30Z": "Débits de boissons",
  "10.71C": "Boulangerie-pâtisserie",
  "10.71D": "Pâtisserie",
  "47.22Z": "Boucherie-charcuterie",
  "47.24Z": "Commerce pain/pâtisserie",
  "56.21Z": "Services traiteurs",
};

const DEPARTEMENTS = ["22", "29", "35", "56"];
const JOURS = 14;

// --- Utilitaires ---

function dateLimite() {
  const d = new Date();
  d.setDate(d.getDate() - JOURS);
  return d.toISOString().split("T")[0];
}

function extraireDirigeant(dirigeants) {
  if (!dirigeants || dirigeants.length === 0) return "";
  const d = dirigeants[0];
  if ((d.type_dirigeant || "").includes("physique")) {
    return `${d.prenoms || ""} ${d.nom || ""}`.trim();
  }
  return d.denomination || "";
}

// --- Récupération API ---

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function fetchAvecRetry(url, tentatives = 3) {
  for (let i = 0; i < tentatives; i++) {
    const resp = await fetch(url);
    if (resp.ok) return resp;
    if (resp.status === 429) {
      const attente = (i + 1) * 2000;
      console.log(`Rate limit (429), attente ${attente}ms avant retry...`);
      await sleep(attente);
    } else {
      throw new Error(`Erreur API : ${resp.status}`);
    }
  }
  throw new Error("Erreur API : trop de tentatives (429)");
}

async function fetchDepartement(departement, depuis) {
  const resultats = [];
  let page = 1;

  while (true) {
    const params = new URLSearchParams({
      activite_principale: APE_CODES.join(","),
      departement,
      etat_administratif: "A",
      date_creation_min: depuis,
      per_page: "25",
      page: String(page),
    });

    const resp = await fetchAvecRetry(`${API_URL}?${params}`);
    const data = await resp.json();
    const results = data.results || [];
    if (results.length === 0) break;

    resultats.push(...results);
    if (page >= (data.total_pages || 1)) break;
    page++;
    await sleep(300);
  }

  return resultats;
}

async function fetchToutesPages(depuis) {
  const resultats = [];
  for (const dept of DEPARTEMENTS) {
    console.log(`Récupération département ${dept}...`);
    resultats.push(...await fetchDepartement(dept, depuis));
    await sleep(500);
  }
  return resultats;
}

// --- Email HTML ---

function construireEmailHtml(entreprises, depuis) {
  const lignes = entreprises
    .map((e) => {
      const siege = e.siege || {};
      return `
        <tr>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">${e.nom_complet || ""}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">${extraireDirigeant(e.dirigeants || [])}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">${siege.libelle_commune || ""}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">${APE_LABELS[e.activite_principale] || e.activite_principale || ""}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">${e.date_creation || ""}</td>
        </tr>`;
    })
    .join("");

  return `
    <div style="font-family:sans-serif;max-width:800px;margin:0 auto;padding:24px;">
      <h2 style="color:#2563eb;">Nouvelles immatriculations — Bretagne</h2>
      <p style="color:#64748b;">${entreprises.length} nouvelle(s) immatriculation(s) détectée(s) depuis le ${depuis}.</p>
      <table style="width:100%;border-collapse:collapse;margin-top:16px;">
        <thead>
          <tr style="background:#f8fafc;">
            <th style="padding:10px 12px;text-align:left;font-size:12px;color:#64748b;text-transform:uppercase;">Raison sociale</th>
            <th style="padding:10px 12px;text-align:left;font-size:12px;color:#64748b;text-transform:uppercase;">Dirigeant</th>
            <th style="padding:10px 12px;text-align:left;font-size:12px;color:#64748b;text-transform:uppercase;">Commune</th>
            <th style="padding:10px 12px;text-align:left;font-size:12px;color:#64748b;text-transform:uppercase;">Activité</th>
            <th style="padding:10px 12px;text-align:left;font-size:12px;color:#64748b;text-transform:uppercase;">Date création</th>
          </tr>
        </thead>
        <tbody>${lignes}</tbody>
      </table>
      <p style="margin-top:24px;color:#64748b;font-size:13px;">
        Consultez le tableau de bord complet sur votre plateforme Netlify.
      </p>
    </div>`;
}

// --- Google Business ---

// --- Handler principal ---

async function run() {
  const supabase = createClient(
    process.env.SUPABASE_URL,
    process.env.SUPABASE_SERVICE_ROLE_KEY
  );

  const depuis = dateLimite();
  console.log(`Recherche depuis le ${depuis}...`);

  const toutes = await fetchToutesPages(depuis);
  console.log(`${toutes.length} nouvelles immatriculations trouvées depuis le ${depuis}.`);

  const retenues = toutes;

  if (retenues.length === 0) {
    console.log("Aucune nouvelle immatriculation sur la période.");
    return { statusCode: 200 };
  }

  // Récupérer les SIRENs déjà connus dans Supabase
  const sirensRetenues = retenues.map((e) => e.siren);
  const { data: dejaDansBdd } = await supabase
    .from("immatriculations")
    .select("siren")
    .in("siren", sirensRetenues);

  const sirensConnus = new Set((dejaDansBdd || []).map((r) => r.siren));
  const vraimentNouveaux = retenues.filter((e) => !sirensConnus.has(e.siren));
  console.log(`${vraimentNouveaux.length} vraiment nouvelle(s) (pas encore en base).`);

  // Upsert dans Supabase (le siren unique évite les doublons)
  const rows = retenues.map((e) => ({
    siren: e.siren,
    raison_sociale: e.nom_complet || "",
    dirigeant: extraireDirigeant(e.dirigeants || []),
    adresse: e.siege?.adresse || "",
    code_postal: e.siege?.code_postal || "",
    commune: e.siege?.libelle_commune || "",
    date_creation: e.date_creation || null,
    code_ape: e.activite_principale || "",
    activite: APE_LABELS[e.activite_principale] || e.activite_principale || "",
  }));

  const { error } = await supabase
    .from("immatriculations")
    .upsert(rows, { onConflict: "siren" });

  if (error) {
    console.error("Erreur Supabase :", error);
    throw error;
  }

  console.log(`${rows.length} entrées sauvegardées dans Supabase.`);

  // Email uniquement si de vraies nouvelles entreprises
  if (vraimentNouveaux.length === 0) {
    console.log("Pas de nouvelle entreprise depuis la dernière vérification.");
    return { statusCode: 200 };
  }

  // Envoi email
  const resend = new Resend(process.env.RESEND_API_KEY);
  const date = new Date().toLocaleDateString("fr-FR", {
    day: "numeric",
    month: "long",
    year: "numeric",
  });

  await resend.emails.send({
    from: "Veille Immatriculations <onboarding@resend.dev>",
    to: process.env.ALERT_EMAIL,
    subject: `🏪 ${vraimentNouveaux.length} nouvelle(s) immatriculation(s) détectée(s) — ${date}`,
    html: construireEmailHtml(vraimentNouveaux, depuis),
  });

  console.log(`Email envoyé à ${process.env.ALERT_EMAIL}.`);
  return { statusCode: 200 };
}

export const handler = schedule("0 8 * * *", run);

File: scripts/check_immatriculations.py
"""
Veille immatriculations Bretagne — script GitHub Actions
Interroge l'API data.gouv, filtre par date, upsert Supabase, alerte Resend.
"""

import os
import time
from datetime import datetime, timedelta

import requests

# --- Config (variables d'environnement GitHub Actions) ---

API_URL = "https://recherche-entreprises.api.gouv.fr/search"
SUPABASE_URL = os.environ["SUPABASE_URL"]
SUPABASE_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
RESEND_API_KEY = os.environ["RESEND_API_KEY"]
ALERT_EMAIL = os.environ["ALERT_EMAIL"]
SERPAPI_KEY = os.environ.get("SERPAPI_KEY", "")

APE_CODES = [
    "56.10A", "56.10C", "55.10Z", "56.30Z",
    "10.71C", "10.71D", "47.22Z", "47.24Z", "56.21Z",
]
APE_LABELS = {
    "56.10A": "Restauration traditionnelle",
    "56.10C": "Restauration rapide",
    "55.10Z": "Hôtels et hébergement",
    "56.30Z": "Débits de boissons",
    "10.71C": "Boulangerie-pâtisserie",
    "10.71D": "Pâtisserie",
    "47.22Z": "Boucherie-charcuterie",
    "47.24Z": "Commerce pain/pâtisserie",
    "56.21Z": "Services traiteurs",
}
DEPARTEMENTS = ["22", "29", "35", "56"]
JOURS = 14


# --- Utilitaires ---

def date_limite():
    return (datetime.now() - timedelta(days=JOURS)).strftime("%Y-%m-%d")


def extraire_dirigeant(dirigeants):
    if not dirigeants:
        return ""
    d = dirigeants[0]
    if "physique" in d.get("type_dirigeant", ""):
        return f"{d.get('prenoms', '')} {d.get('nom', '')}".strip()
    return d.get("denomination", "")


# --- API data.gouv ---

def fetch_avec_retry(url, params, tentatives=3):
    for i in range(tentatives):
        r = requests.get(url, params=params, timeout=30)
        if r.ok:
            return r
        if r.status_code == 429:
            attente = (i + 1) * 2
            print(f"  Rate limit 429, attente {attente}s...")
            time.sleep(attente)
        else:
            raise Exception(f"Erreur API : {r.status_code}")
    raise Exception("Trop de tentatives (429)")


def fetch_departement(departement):
    resultats = []
    page = 1
    while True:
        params = {
            "activite_principale": ",".join(APE_CODES),
            "departement": departement,
            "etat_administratif": "A",
            "per_page": 25,
            "page": page,
        }
        r = fetch_avec_retry(API_URL, params)
        data = r.json()
        results = data.get("results", [])
        if not results:
            break
        resultats.extend(results)
        if page >= data.get("total_pages", 1):
            break
        page += 1
        time.sleep(0.3)
    return resultats


def fetch_toutes_pages():
    resultats = []
    for dept in DEPARTEMENTS:
        print(f"  Département {dept}...")
        resultats.extend(fetch_departement(dept))
        time.sleep(0.5)
    return resultats


def filtrer(entreprises, depuis):
    return [e for e in entreprises if (e.get("date_creation") or "") >= depuis]


# --- Supabase REST API ---

def supabase_headers():
    return {
        "apikey": SUPABASE_KEY,
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
    }


def get_sirens_connus(sirens):
    if not sirens:
        return set()
    r = requests.get(
        f"{SUPABASE_URL}/rest/v1/immatriculations",
        params={"select": "siren", "siren": f"in.({','.join(sirens)})"},
        headers=supabase_headers(),
        timeout=30,
    )
    r.raise_for_status()
    return {row["siren"] for row in r.json()}


def upsert_supabase(rows):
    r = requests.post(
        f"{SUPABASE_URL}/rest/v1/immatriculations?on_conflict=siren",
        json=rows,
        headers={**supabase_headers(), "Prefer": "resolution=merge-duplicates,return=minimal"},
        timeout=60,
    )
    r.raise_for_status()


# --- Scan Google Business ---

def _nom_similaire(nom_cherche, nom_trouve):
    n1 = nom_cherche.lower().strip()
    n2 = nom_trouve.lower().strip()
    mots = [m for m in n1.split() if len(m) > 2]
    return any(mot in n2 for mot in mots)


def _commune_similaire(commune, adresse):
    return commune.lower().strip() in (adresse or "").lower()


def check_google_business(nom, commune):
    r = requests.get("https://serpapi.com/search.json", params={
        "engine": "google_maps",
        "q": f"{nom} {commune}",
        "hl": "fr",
        "gl": "fr",
        "api_key": SERPAPI_KEY,
    }, timeout=15)
    if not r.ok:
        return {"google_business": None}

    results = r.json().get("local_results", [])
    vide = {"google_business": False, "google_place_id": None, "google_rating": None,
            "google_reviews": None, "phone": None, "website": None}
    if not results:
        return vide

    first = results[0]
    if not _nom_similaire(nom, first.get("title", "")) or \
       not _commune_similaire(commune, first.get("address", "")):
        return vide

    return {
        "google_business": True,
        "google_place_id": first.get("place_id"),
        "google_rating": first.get("rating"),
        "google_reviews": first.get("reviews"),
        "phone": first.get("phone"),
        "website": first.get("website"),
    }


def get_non_scannes():
    r = requests.get(
        f"{SUPABASE_URL}/rest/v1/immatriculations",
        params={"select": "siren,raison_sociale,commune", "google_checked_at": "is.null"},
        headers=supabase_headers(),
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def update_google(siren, data):
    data["google_checked_at"] = datetime.now().astimezone().isoformat()
    r = requests.patch(
        f"{SUPABASE_URL}/rest/v1/immatriculations",
        json=data,
        params={"siren": f"eq.{siren}"},
        headers=supabase_headers(),
        timeout=15,
    )
    r.raise_for_status()


def scan_google():
    if not SERPAPI_KEY:
        print("SERPAPI_KEY non défini, scan Google ignoré.")
        return

    entreprises = get_non_scannes()
    print(f"{len(entreprises)} entreprise(s) à scanner sur Google Business...")

    for i, e in enumerate(entreprises, 1):
        nom = e.get("raison_sociale", "")
        commune = e.get("commune", "")
        siren = e.get("siren", "")
        print(f"  [{i}/{len(entreprises)}] {nom} ({commune})...", end=" ", flush=True)

        google = check_google_business(nom, commune)
        update_google(siren, google)

        statut = "✅" if google.get("google_business") else "❌"
        print(f"{statut} Tel: {google.get('phone') or '-'}")
        time.sleep(1)

    print("Scan Google terminé.")


# --- Email Resend ---

def construire_html(entreprises, depuis):
    lignes = "".join(f"""
        <tr>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">{e.get("nom_complet","")}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">{extraire_dirigeant(e.get("dirigeants",[]))}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">{e.get("siege",{}).get("libelle_commune","")}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">{APE_LABELS.get(e.get("activite_principale",""), e.get("activite_principale",""))}</td>
          <td style="padding:10px 12px;border-bottom:1px solid #e2e8f0;">{e.get("date_creation","")}</td>
        </tr>""" for e in entreprises)

    return f"""
    <div style="font-family:sans-serif;max-width:800px;margin:0 auto;padding:24px;">
      <h2 style="color:#2563eb;">Nouvelles immatriculations — Bretagne</h2>
      <p style="color:#64748b;">{len(entreprises)} nouvelle(s) immatriculation(s) depuis le {depuis}.</p>
      <table style="width:100%;border-collapse:collapse;margin-top:16px;">
        <thead><tr style="background:#f8fafc;">
          <th style="padding:10px;text-align:left;font-size:12px;color:#64748b;">Raison sociale</th>
          <th style="padding:10px;text-align:left;font-size:12px;color:#64748b;">Dirigeant</th>
          <th style="padding:10px;text-align:left;font-size:12px;color:#64748b;">Commune</th>
          <th style="padding:10px;text-align:left;font-size:12px;color:#64748b;">Activité</th>
          <th style="padding:10px;text-align:left;font-size:12px;color:#64748b;">Date création</th>
        </tr></thead>
        <tbody>{lignes}</tbody>
      </table>
      <p style="margin-top:24px;color:#64748b;font-size:13px;">
        Tableau de bord : <a href="https://immatriculations-22.vercel.app">immatriculations-22.vercel.app</a>
      </p>
    </div>"""


def envoyer_email(entreprises, depuis):
    date = datetime.now().strftime("%-d %B %Y")
    r = requests.post(
        "https://api.resend.com/emails",
        json={
            "from": "Veille Immatriculations <onboarding@resend.dev>",
            "to": ALERT_EMAIL,
            "subject": f"🏪 {len(entreprises)} nouvelle(s) immatriculation(s) — {date}",
            "html": construire_html(entreprises, depuis),
        },
        headers={"Authorization": f"Bearer {RESEND_API_KEY}", "Content-Type": "application/json"},
        timeout=30,
    )
    r.raise_for_status()
    print(f"  Email envoyé à {ALERT_EMAIL}.")


# --- Point d'entrée ---

def main():
    depuis = date_limite()
    print(f"Recherche depuis le {depuis}...")

    toutes = fetch_toutes_pages()
    print(f"{len(toutes)} entreprises actives trouvées.")

    retenues = filtrer(toutes, depuis)
    print(f"{len(retenues)} nouvelles immatriculations.")

    if not retenues:
        print("Aucune nouvelle immatriculation.")
        return

    sirens = [e["siren"] for e in retenues if e.get("siren")]
    sirens_connus = get_sirens_connus(sirens)
    vraiment_nouveaux = [e for e in retenues if e.get("siren") and e["siren"] not in sirens_connus]
    print(f"{len(vraiment_nouveaux)} vraiment nouvelle(s) (pas encore en base).")

    rows = [{
        "siren": e.get("siren", ""),
        "raison_sociale": e.get("nom_complet", ""),
        "dirigeant": extraire_dirigeant(e.get("dirigeants", [])),
        "adresse": e.get("siege", {}).get("adresse", ""),
        "code_postal": e.get("siege", {}).get("code_postal", ""),
        "commune": e.get("siege", {}).get("libelle_commune", ""),
        "date_creation": e.get("date_creation") or None,
        "code_ape": e.get("activite_principale", ""),
        "activite": APE_LABELS.get(e.get("activite_principale", ""), e.get("activite_principale", "")),
    } for e in retenues]

    upsert_supabase(rows)
    print(f"{len(rows)} entrées sauvegardées dans Supabase.")

    if vraiment_nouveaux:
        envoyer_email(vraiment_nouveaux, depuis)

    scan_google()


if __name__ == "__main__":
    main()
